src/backend/error.rs

```diff
@@ -0,0 +1,83 @@
+use std::{error::Error, fmt::Display};
+
+use super::Context;
+
+pub(crate) fn format_error(err: &(dyn Error + 'static), context: &Context) -> String {
+    let mut err_descs = Vec::new();
+
+    err_descs.push(fmt_pretty_any(err, context));
+
+    if err.source().is_some() {
+        let mut source_opt = err.source();
+        while let Some(source) = source_opt {
+            err_descs.push(fmt_pretty_any(source, context));
+            source_opt = source.source();
+        }
+    }
+
+    let desc = format!("Validation Error\n\nCaused by:\n{}", err_descs.join(""));
+    desc
+}
+
+fn fmt_pretty_any(error: &(dyn Error + 'static), context: &Context) -> String {
+    if let Some(pretty_err) = error.downcast_ref::<wgc::command::TransferError>() {
+        pretty_err.fmt_pretty(context)
+    } else {
+        format_error_line(error.as_display())
+    }
+}
+
+fn format_error_line(err: &dyn Display) -> String {
```

Comment:
these should probably work with `fmt::Arguments` parameter instead of `&dyn Display`, see https://doc.rust-lang.org/std/fmt/struct.Arguments.html